Small improvements

davidegrohmann opened the pull request:
Builds on #8568

davidegrohmann committed:
Small improvements

davidegrohmann committed:
Do not load whole node records for simply check inUse flag

When fetching all nodes there is no need to deserialise records for
simply throwing them away.

davidegrohmann committed:
Do not retrieve the record if not inUse when visiting a relationship

davidegrohmann committed:
Simplify NodeItem interface

Add new operations to RawCursor

Simlify a lot of other code as well in the process

davidegrohmann committed:
Fix loop forever in TxSingleNodeCursor

chrisvest commented:
@davidegrohmann I've merged #8568

davidegrohmann committed:
Flatten node cursors

davidegrohmann commented:
@chrisvest rebased

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1006,15 +1006,10 @@ public RECORD newRecord()
-        // on that record, so it's to ensure it isn't forgotten.
-        record.setId( id );
         long pageId = pageIdForRecord( id );
-        int offset = offsetForId( id );
```
The reason these two were kept together was because they have the same formula, except page id uses divide and offset uses modulo. At the assembly level, the division instruction computes both the division and the modulo at the same time, so by keeping the two together, the optimiser can do both calls with a single division instruction. This is a useful optimisation because division is pretty expensive.

You can actually still keep them together if you move both into the `readIntoRecord` method, and pass `0` as the page id to `storeFile.io()`. This works because `readIntoRecord` calls `cursor.next( pageId )` anyway, so the page id passed to `io` isn't used.

davidegrohmann replied:
I'll do that

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/CommonAbstractStoreBehaviourTest.java:
```diff
@@ -257,6 +258,60 @@ public void pageCursorErrorsMustNotLingerInRecordCursor() throws Exception
+        // next with id does not affect the old pointer either, so 42 is read now
+        assertTrue( cursor.next() );
+        assertEquals( record42, cursor.get() );
+    }
```
👍

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/AllRelationshipIterator.java:
```diff
@@ -38,4 +44,25 @@
+                if ( record.inUse() )
+                {
+                    return next( record.getId() );
+                }
```
Should this not also be implemented in terms of `store.isInUse( currentId )` ?

davidegrohmann replied:
That would be nice, but you need the record in `relationshipVisit` which is called on each element in the iterator.

davidegrohmann committed:
Simplify StoreSingleNodeCursor

Extracting common code patterns into Cursor interfaces
Extract inner classes
Share common code among different classes to avoid duplication

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/txstate/TransactionCountingStateVisitor.java:
```diff
@@ -68,35 +69,27 @@ public void visitCreatedNode( long id )
-        }
+            } );
+
+        } );
```
Stylistically, this might be more approachable if the outer-most lambda was turned into a reference to a private helper method.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/txstate/TransactionCountingStateVisitor.java:
```diff
@@ -141,30 +133,18 @@ public void visitNodeLabelChanges( long id, final Set<Integer> added, final Set<
+                    updateRelationshipsCountsFromDegrees( degree.type(), label, -degree.outgoing(),
+                            -degree.incoming() );
+                }
+            } ) );
```
Same here, where two lambda clauses are started on the same line.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/TwoPhaseNodeForRelationshipLocking.java:
```diff
@@ -26,104 +26,104 @@
     {
+        boolean retry = true;
         nodeIds.add( nodeId );
         while ( retry )
```
This could be a do-while, I think.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/TwoPhaseNodeForRelationshipLocking.java:
```diff
@@ -26,104 +26,104 @@
-                        break;
+                        retry = performAction( state, relationships.get(), first );
+                        first = false;
                     }
```
Very nice. These changes makes it much easier to see what's going on.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/TwoPhaseNodeForRelationshipLockingTest.java:
```diff
@@ -204,7 +206,42 @@ public long next()
+                throw new EntityNotFoundException( EntityType.NODE, nodeId );
+            }
+            return cursor;
+        } );
```
Too much logic in these mocks :( These tests should have been written with stubs, and in terms of observable behaviour from the start.

davidegrohmann committed:
Introduce RelationshipTypeItem instead of using a generic IntSupplier

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/LabelTransactionStateTest.java:
```diff
@@ -362,23 +356,21 @@ private void commitLabels( Integer... labels ) throws Exception
-            }
-        }
+        txContext.nodeCursorById( state, nodeId ).forAll( node -> assertEquals( asSet( labels ),
+                node.labels().mapReduce( new HashSet<>(), IntSupplier::getAsInt, this::addToCollection ) ) );
```
Maybe we want a `collect` method to simplify these uses. Looks like this has come up a couple of times now.

Is it intentional that you are now no longer collecting into a primitive collection?

davidegrohmann replied:
Yeah maybe a `collect` makes sense.

We can use a primitive collection but I didn't think it was important in a test.

chrisvest commented on community/primitive-collections/src/main/java/org/neo4j/cursor/RawCursor.java:
```diff
@@ -50,11 +52,33 @@
+    }
+
+    default <E> void mapForAll( Function<T,E> function, Consumer<E> consumer ) throws EXCEPTION
+    {
```
I don't know if you have any long term plans for this interface, but if we implemented `Spliterator` for `RawCursor`, we could give it a `stream()` method, and access the power of the streams framework.

davidegrohmann replied:
I didn't explore that possibility. I might give it a try.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -380,15 +383,40 @@ public RelationshipIterator nodeGetRelationships( long nodeId, Direction directi
+            types = Arrays.copyOf( types, unique );
+        }
+        return types;
+    }
```
We could alternatively sort the types, and do the uniqueness check in a single pass. _Iff_ it could help us to know that the types are sorted up front, then this would be a good idea. I haven't checked if this premise is true, but it sounds like a logical thing.

davidegrohmann replied:
Yep that might an improvement, I simply moved that code as is from another class. No changes to the logic there.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DegreeCounter.java:
```diff
@@ -0,0 +1,104 @@
+        default:
+            throw new IllegalArgumentException( direction.name() );
+        }
+    }
```
Maybe you want these methods to be static instead of default? They don't really rely on any methods in this or super interfaces. Default methods are primarily for adding methods to existing interfaces in a backwards compatible way. They're also harder for the JIT to optimise, because of their complicated resolution logic.

davidegrohmann replied:
I'll do that.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DegreeItemCursor.java:
```diff
@@ -0,0 +1,92 @@
+    {
+        if ( keys == null )
+        {
+            throw new IllegalStateException();
```
Perhaps mention that the cursor has been accessed after it's been closed.

davidegrohmann replied:
sure

davidegrohmann replied:
Actually key is set to null also when a next value cannot be found.

chrisvest reviewed (commented):
Please take a look at the comments I have.

davidegrohmann committed:
Keep together the calls to pageIdForRecord and offsetForId

davidegrohmann committed:
Change to do-while loop

Fix initialization of the set in the while loop

davidegrohmann committed:
Make Degree counter a class with static methods

davidegrohmann committed:
Add error message to exception

davidegrohmann committed:
Add collect to RawCursor

Simplify code

davidegrohmann committed:
Extract RecordCursor from CommonAbstractStore
